Proper URI rendering for Zend/Http

afeder opened the pull request:
Fixes #7472.

Martin-P commented:
This PR needs unittests.

afeder commented:
Sorry, what is a "PR"? Does any guidance exist as to what kinds of unit tests would be needed?

samsonasik commented:
need rebase

afeder commented:
samsonasik: Indeed. Done.

afeder commented:
I added a test for the issue described in #7472.

afeder committed:
Proper request URI rendering for Zend/Http/Request and Zend/Http/Client.

samsonasik commented:
need rebase again

afeder commented:
Rebase done.

sasezaki commented on library/Zend/Http/Request.php:
```diff
@@ -521,13 +580,60 @@ public function isFlashRequest()
+    /**
+     * Return the formatted URI for this http request
+     *
+     * @return string
```
`@return HttpUri` , not ?

Martin-P commented:
IMO this needs PR needs many more tests.

There are multiple new methods introduced in `Zend\Http\Request` which are not tested at all:
- `setOptions`
- `setArgSeparator`
- `getArgSeparator`
- `renderQueryString`
- `renderUri`

A big chunk of code in `Zend\Http\Client` has been removed and new code is added, but no tests have been added for this.

The only test which has been added is a very basic query string example. What happens with more advanced query strings? Does that work too or will it fail?

afeder commented:
If you share with me what tests you are looking for I shall add them right away.

The big chunks of code being removed or added is little more than migrating the URI building logic from the Client class into the Request class.

I've found two tests pertaining to this code which I've now copied over to the tests for the Request class. I've also expanded the query test to cover the case of multiple query parameters being set, and duplicated this test for the renderUri() and renderQueryString() functions.

sasezaki commented on library/Zend/Http/Request.php:
```diff
@@ -166,6 +176,55 @@ public static function fromString($string, $allowCustomMethods = true)
+            $options = ArrayUtils::iteratorToArray($options);
+        }
+        if (!is_array($options)) {
+            throw new Request\Exception\InvalidArgumentException('Config parameter is not valid');
```
should be `Exception\InvalidArgumentException`, please add test too.

afeder committed:
Fix setOptions() exception and add test per sasezaki's comment.

afeder commented:
I fixed the exception in the setOptions() function per sasezaki's comment. As for test, none exist in the original code in Client.php, but I added a basic one anyway.

GeeH commented:
This issue has been moved from the `zendframework` repository as part of the bug migration program as outlined here - http://framework.zend.com/blog/2016-04-11-issue-closures.html
New issue can be found at: https://github.com/zendframework/zend-http/issues/68
